File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer>arr = new ArrayList<Integer>();
        for(int i=0; i<nums.length; i++) {
            if(arr.isEmpty() || arr.get(arr.size() - 1) < nums[i]) {
                arr.add(nums[i]);
            }
            else {
                int ind = Collections.binarySearch(arr, nums[i]);
                if(ind < 0) ind = - (ind + 1);
                arr.set(ind, nums[i]);
            }
        }
        return (int) arr.size();
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int>ans;
        for(int i=0; i<nums.size(); i++) {
            if(ans.empty() || ans.back() < nums[i]) {
                ans.push_back(nums[i]);
            }
            else {
                int ind = lower_bound(ans.begin(), ans.end(), nums[i]) - ans.begin();
                ans[ind] = nums[i];
            }
        }
        return ans.size();
    }
};